### PR TITLE
Add P2PInterface and P2PMsgHandler

### PR DESCRIPTION
**Pull request by chaychen2005**

### CODE REVIEW

File: libp2p/Common.h [hunk 1]
```diff
@@ -105,6 +105,66 @@ enum DisconnectReason
     NoDisconnect = 0xffff
 };
 
+DEV_SIMPLE_EXCEPTION(NetworkTimeout);
+DEV_SIMPLE_EXCEPTION(NetworkUnreachable);
+DEV_SIMPLE_EXCEPTION(TopicNotFound);
+DEV_SIMPLE_EXCEPTION(NodeIDNotFound);
+DEV_SIMPLE_EXCEPTION(OK);
```
Comment (wheatli) [inline libp2p/Common.h]: OK exception?

File: libp2p/Common.h [hunk 2]
```diff
@@ -105,6 +105,66 @@ enum DisconnectReason
...
+    {
+        if (isRequestPacket())
+            return -m_protocolID;
+        else
+            return 0;
+    }
+
+private:
+    uint32_t m_length = 0;     ///< m_length = HEADER_LENGTH + length(m_buffer)
+    int32_t m_protocolID = 0;  ///< message type
```
Comment (wheatli) [inline libp2p/Common.h]: Why not use uint32_t

File: libp2p/P2PMsgHandler.cpp [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+ */
+
+#include "P2PMsgHandler.h"
+
+namespace dev
+{
+namespace p2p
+{
+bool P2PMsgHandler::addSeq2Callback(
+    uint32_t seq, std::function<void(dev::Exception, Message::Ptr)> callback)
```
Comment (wheatli) [inline libp2p/P2PMsgHandler.cpp]: callback should be const&

File: libp2p/P2PMsgHandler.cpp [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+    }
+    else
+    {
+        LOG(ERROR) << "Handler not found! SeqID = " << seq;
+        return false;
+    }
+}
+
+bool P2PMsgHandler::addProtocolID2Handler(uint32_t protocolID,
+    std::function<void(dev::Exception, std::shared_ptr<Session>, Message::Ptr)> callback)
```
Comment (wheatli) [inline libp2p/P2PMsgHandler.cpp]: callback should be const&, plz check all addXX function

File: libp2p/Service.h [hunk 5]
```diff
@@ -34,92 +35,78 @@ namespace dev
 {
 namespace p2p
 {
-DEV_SIMPLE_EXCEPTION(NetworkTimeout);
-DEV_SIMPLE_EXCEPTION(NetworkUnreachable);
-DEV_SIMPLE_EXCEPTION(TopicNotFound);
-DEV_SIMPLE_EXCEPTION(NodeIDNotFound);
-
-class Message : std::enable_shared_from_this<Message>
+class P2PInterfase
```
Comment (wheatli) [inline libp2p/Service.h]: place it as a independent file

File: libp2p/Common.h (hunk 1 above)
Comment (chaychen2005) [reply libp2p/Common.h]: This problem is also found in the use of exceptions. Now add an exception class under P2P, with a field indicating whether it is successful.

File: libp2p/Common.h (hunk 2 above)
Comment (chaychen2005) [reply libp2p/Common.h]: int32_t indicates the possibility of positive or negative Numbers. A positive number indicates the sending package and a negative number indicates the response package.

File: libp2p/P2PMsgHandler.cpp (hunk 3 above)
Comment (chaychen2005) [reply libp2p/P2PMsgHandler.cpp]: I see what you mean.

File: libp2p/P2PMsgHandler.cpp (hunk 4 above)
Comment (chaychen2005) [reply libp2p/P2PMsgHandler.cpp]: I see what you mean.

File: libp2p/Service.h (hunk 5 above)
Comment (chaychen2005) [reply libp2p/Service.h]: I see what you mean , and this code in a new P2PInterface.h in next pr.

File: libp2p/Service.h [hunk 6]
```diff
@@ -34,92 +36,49 @@ namespace dev
...
 public:
-    typedef std::shared_ptr<Message> Ptr;
-
-    const static size_t HEADER_LENGTH = sizeof(uint32_t) + sizeof(uint32_t) +
-                                        sizeof(uint32_t);  ///< HEADER_LENGTH = length(m_length) +
-                                                           ///< length(m_protocolID) + length(m_seq)
-    const static size_t MAX_LENGTH = 1024 * 1024;          ///< The maximum length of data is 1M.
-
-    Message() { m_buffer = std::make_shared<bytes>(); }
+    Service(std::shared_ptr<Host> _host) : m_host(_host)
```
Comment (wheatli) [inline libp2p/Service.h]: I think its better to pass in the p2pMsgHandler, you will get it when you write unittest

File: libp2p/Service.h [hunk 7]
```diff
@@ -34,92 +36,49 @@ namespace dev
...
+    ~Service() { m_ioService = NULL; }
 
-    uint32_t length() { return m_length; }
-    void setLength(uint32_t _length) { m_length = _length; }
+    Message::Ptr sendMessageByNodeID(NodeID const& nodeID, Message::Ptr message) override;
 
-    uint32_t protocolID() { return m_protocolID; }
-    void setProtocolID(uint32_t _protocolID) { m_protocolID = _protocolID; }
+    void asyncSendMessageByNodeID(NodeID const& nodeID, Message::Ptr message,
+        std::function<void(P2PException, Message::Ptr)> callback, Options const& options) override;
```
Comment (wheatli) [inline libp2p/Service.h]: I have seen so many std::function<void(P2PException, Message::Ptr)>, you can declare a variable as CallbackFunc=std::function<void(P2PException, Message::Ptr)>

File: libp2p/Session.cpp [hunk 8]
```diff
@@ -416,23 +305,72 @@ bool Session::checkRead(std::size_t _expected, boost::system::error_code _ec, st
 
         return false;
     }
-    else if (_ec && _length < _expected)
+
+    return true;
+}
+
+void Session::onMessage(P2PException e, std::shared_ptr<Session> session, Message::Ptr message)
```
Comment (wheatli) [inline libp2p/Session.cpp]: e should be const&

File: libp2p/Session.cpp [hunk 9]
```diff
@@ -416,23 +305,72 @@ bool Session::checkRead(std::size_t _expected, boost::system::error_code _ec, st
...
+            if (callback->callbackFunc)
+            {
+                LOG(INFO) << "Session::onMessage, call callbackFunc by seq=" << message->seq();
+                ///< TODO: use threadPool
+                callback->callbackFunc(e, message);
+            }
+        }
+        else
+        {
+            LOG(ERROR) << "Session::onMessage, callback not found by seq=" << message->seq();
```
Comment (wheatli) [inline libp2p/Session.cpp]: Sometimes I don't need a callback, I just send a msg to someone without any hope to recv response...How do you adapt my situation?

Comment (chaychen2005) [reply libp2p/Session.cpp]: Someone can write a callback function got by protocolID without sending response package. It is up to someone.

File: libp2p/Service.cpp [hunk 10]
```diff
@@ -20,22 +20,183 @@
...
-    dev::h256 const& nodeID, uint32_t protocolID, Message::Ptr message)
+Message::Ptr Service::sendMessageByNodeID(NodeID const& nodeID, Message::Ptr message)
 {
+    LOG(INFO) << "Call Service::sendMessageByNodeID";
+    try
+    {
+        bool findNodeID = false;
+        {
+            RecursiveGuard l(m_host->mutexSessions());
+            std::unordered_map<NodeID, std::shared_ptr<SessionFace>> s = m_host->sessions();
```
[the remark this answers is not in the document]
Comment (chaychen2005) [reply libp2p/Service.cpp]: Unified adjust later.

File: libp2p/Service.cpp [hunk 11]
```diff
@@ -20,22 +20,183 @@
...
     return Message::Ptr();
 }
 
-void Service::asyncSendMessageByNodeID(dev::h256 const& nodeID, uint32_t protocolID,
-    Message::Ptr message, std::function<void(dev::Exception, Message::Ptr)> callback,
-    Options options)
+void Service::asyncSendMessageByNodeID(NodeID const& nodeID, Message::Ptr message,
+    std::function<void(P2PException, Message::Ptr)> callback, Options const& options)
+{
+    LOG(INFO) << "Call Service::asyncSendMessageByNodeID to NodeID=" << toJS(nodeID);
```
[the remark this answers is not in the document]
Comment (chaychen2005) [reply libp2p/Service.cpp]: Unified adjust later.

File: libp2p/Session.cpp (hunk 9 above)
Comment (chaychen2005) [reply libp2p/Session.cpp]: Now add a function to deal with sending message without callback.

File: libp2p/Common.h [hunk 12]
```diff
@@ -191,59 +191,45 @@ class P2PException : public std::exception
...
 };
 
+#define CallbackFunc std::function<void(P2PException, Message::Ptr)>
 struct ResponseCallback : public std::enable_shared_from_this<ResponseCallback>
 {
     typedef std::shared_ptr<ResponseCallback> Ptr;
 
-    std::function<void(P2PException, Message::Ptr)> callbackFunc;
-    std::shared_ptr<boost::asio::deadline_timer> timeoutHandler;
+    CallbackFunc m_callbackFunc;
```
[the remark this answers is not in the document]
Comment (chaychen2005) [reply libp2p/Common.h]: Done. Fix in commit c27d768, and now there is not 'm_' in struct.

**Pull request merged**